onnxruntime/core/graph/contrib_ops/contrib_defs.cc

```diff
@@ -2679,6 +2679,140 @@ It's an extension of Gelu. It takes the sum of input A and bias input B as the i
 
+  static const char* OptionalConstruct_ver1_doc = R"DOC(
+      Construct an optional type containing either an empty optional of a certain type specified by the attribute, "
+      "or an optional type containing the 'input' element."
+      )DOC";
+
+  ONNX_CONTRIB_OPERATOR_SCHEMA(OptionalConstruct)
+      .SetDomain(kMSDomain)
+      .SinceVersion(1)
+      .SetDoc(OptionalConstruct_ver1_doc)
+      .Input(0, "input", "The input element.", "T", OpSchema::Optional)
+      .Attr("type", "Type of the element in the optional output", AttributeProto::TYPE_PROTO, OPTIONAL_VALUE)
+      .Output(0, "output", "The optional output enclosing the input element.", "O")
+      .TypeConstraint(
+          "T",
+          {"tensor(float)",
+           "seq(tensor(float))"},
+          "Constrains input type to all tensor and sequence types.")
+      .TypeConstraint(
+          "O",
+          {"optional(tensor(float))", "optional(tensor(uint8))", "optional(tensor(uint16))", 
+           "optional(tensor(uint32))", "optional(tensor(uint64))", "optional(tensor(int8))", 
+           "optional(tensor(int16))", "optional(tensor(int32))", "optional(tensor(int64))", 
+           "optional(tensor(float16))", "optional(tensor(float))", "optional(tensor(double))", 
+           "optional(tensor(string))", "optional(tensor(bool))", "optional(tensor(complex64))", 
+           "optional(tensor(complex128))",
+           "optional(seq(tensor(float)))", "optional(seq(tensor(uint8)))", "optional(seq(tensor(uint16)))", 
+           "optional(seq(tensor(uint32)))", "optional(seq(tensor(uint64)))", "optional(seq(tensor(int8)))", 
+           "optional(seq(tensor(int16)))", "optional(seq(tensor(int32)))", "optional(seq(tensor(int64)))", 
+           "optional(seq(tensor(float16)))", "optional(seq(tensor(float)))", "optional(seq(tensor(double)))", 
+           "optional(seq(tensor(string)))", "optional(seq(tensor(bool)))", "optional(seq(tensor(complex64)))", 
+           "optional(seq(tensor(complex128)))"},
+          "Constrains output type to all optional tensor or optional sequence types.")
+      .TypeAndShapeInferenceFunction([](InferenceContext& ctx) {
+          const size_t numOutputs = ctx.getNumOutputs();
+          if (numOutputs != 1) {
+            fail_type_inference("OptionalConstruct is expected to have an output.");
+          }
+
+          const size_t numInputs = ctx.getNumInputs();
+          const auto* attr_proto = ctx.getAttribute("type");
+
+          if ((numInputs == 0) && (attr_proto != nullptr)) {
+            if (!attr_proto->has_tp())
+              fail_type_inference(
+                  "Attribute 'type' should be a TypeProto and it should specify a type.");
+            auto attr_tp = attr_proto->tp();
+            ctx.getOutputType(0)
+                ->mutable_optional_type()
+                ->mutable_elem_type()
+                ->CopyFrom(attr_tp);
+          } else if (numInputs == 1) {
+            auto input_type = ctx.getInputType(0);
```

Comment:
should it be an error if num inputs is 1 and attr_proto is not null ?

Reply:
Yes, only one of the two (either attr or input) should be provided. I have separated the use-cases so the two won't be confused. Also to prevent mismatches.